src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -539,6 +553,11 @@ class FastSync(
     private def insertBlocks(requestedHashes: Seq[ByteString], blockBodies: Seq[BlockBody]): Unit = {
       (requestedHashes zip blockBodies).foreach { case (hash, body) =>
         blockchain.save(hash, body)
+        blockSyncCache.get(hash, () => (
```

Comment:
(comment is in this line but it refers to every `blockSyncCache.get`)
I am not sure if this is good usage of this method, as docs say this methods basicly replaces idiom `if cached, return; otherwise create, cache and return`, so if you already cached data under header.hash (and you most probably did as headers are processed always before matching bodies) this method will return `(Some(header), false, false))` and not update value under hash. Therefore when the times come to `updateBestBlockIfNeeded` 
you can already saved bodies and reciepts but in cache you will have  `(Some(header), false, false))` and wont update best block. 
Basically best block will be updated only when headers are already out of cache (due to limit ) and receipts are already saved.
This all can lead to times when best block is not updating. Example:
```
12:25:13 Block: 528/6014772. Peers waiting_for_response/connected: 1/3 (0 blacklisted). State: 0/0 nodes.
12:25:16 Received 72 block bodies in 8522 ms
12:25:16 Received 20 receipts in 98 ms
12:25:16 Received 60 receipts in 277 ms
12:25:18 Received 72 block bodies in 218 ms
12:25:18 *** Received 200 block headers in 469 ms ***
12:25:18 Received 60 receipts in 176 ms
12:25:18 Received 60 receipts in 266 ms
12:25:23 Received 128 block bodies in 6848 ms
12:25:23 Received 20 receipts in 110 ms
12:25:23 Received 60 receipts in 260 ms
12:25:28 Received 128 block bodies in 4700 ms
12:25:28 Received 72 block bodies in 230 ms
12:25:28 *** Received 200 block headers in 374 ms ***
12:25:28 Received 60 receipts in 184 ms
12:25:28 Received 60 receipts in 212 ms
12:25:28 Received 60 receipts in 230 ms
12:25:38 Received 20 receipts in 141 ms
12:25:38 Received 128 block bodies in 397 ms
12:25:38 Received 72 block bodies in 718 ms
12:25:38 *** Received 200 block headers in 343 ms ***
12:25:39 Received 60 receipts in 231 ms
12:25:39 Received 60 receipts in 830 ms
12:25:39 Received 20 receipts in 94 ms
12:25:40 Received 128 block bodies in 1010 ms
12:25:40 Received 72 block bodies in 194 ms
12:25:41 Received 60 receipts in 2363 ms
12:25:41 *** Received 200 block headers in 1041 ms ***
12:25:41 Received 60 receipts in 162 ms
12:25:41 Received 60 receipts in 192 ms
12:25:41 Received 60 receipts in 198 ms
12:25:43 Block: 528/6014772. Peers waiting_for_response/connected: 0/3 (0 blacklisted). State: 0/0 nodes.
```

And example when we got bodies with invalidated cache, note how we updating block from `528` to `11401`:
```
12:25:43 Block: 528/6014772. Peers waiting_for_response/connected: 0/3 (0 blacklisted). State: 0/0 nodes.
12:25:48 Received 20 receipts in 119 ms
12:25:53 Received 72 block bodies in 5637 ms
12:25:54 *** Received 200 block headers in 332 ms ***
12:25:54 Received 60 receipts in 145 ms
12:25:54 Received 60 receipts in 251 ms
12:25:58 Received 20 receipts in 223 ms
12:26:03 Received 60 receipts in 5283 ms
12:26:03 Received 128 block bodies in 459 ms
12:26:04 *** Received 200 block headers in 308 ms ***
12:26:04 Received 60 receipts in 243 ms
12:26:08 Received 60 receipts in 258 ms
12:26:11 Received 72 block bodies in 8028 ms
12:26:11 Received 128 block bodies in 23344 ms
12:26:11 Got whole header for block with number 11401
12:26:11 Got whole header for block with number 11402
```

Reply:
You have a right. I will fix that issue